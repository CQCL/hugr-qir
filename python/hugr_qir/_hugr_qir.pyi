def cli(*args: str) -> None: ...
def opt_level_choices() -> list[str]: ...
def compile_target_choices() -> list[str]: ...
def output_format_choices() -> list[str]: ...
