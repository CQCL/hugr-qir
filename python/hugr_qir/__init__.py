def main_cli():
    from ._hugr_qir import cli
    import sys
    cli(*sys.argv)
