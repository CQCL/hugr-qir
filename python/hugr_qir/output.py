from base64 import b64encode
from enum import Enum
from pathlib import Path

from llvmlite.binding import (  # type: ignore[import-untyped]
    create_context,
    parse_assembly,
)


class OutputFormat(Enum):
    LLVMIR = "llvmir"
    BITCODE = "bitcode"
    BASE64 = "base64"


def get_write_mode(out_format: OutputFormat | None, file_path: Path | None) -> str:
    if out_format and out_format in [OutputFormat.BITCODE, OutputFormat.BASE64]:
        return "wb"
    if file_path and file_path.suffix not in [".ll", ".asm"]:
        return "wb"
    return "w"


def ir_string_to_output_format(qir_ir: str, output_format: OutputFormat) -> str | bytes:
    match output_format:
        case OutputFormat.LLVMIR:
            return qir_ir
        case OutputFormat.BITCODE:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            return module.as_bitcode()
        case OutputFormat.BASE64:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            qir_bitcode = module.as_bitcode()
            return b64encode(qir_bitcode).decode("utf-8")
    errmsg = "Unrecognized output format"
    raise ValueError(errmsg)
