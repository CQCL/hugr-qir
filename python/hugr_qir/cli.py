"""Cli for hugr-qir."""

import tempfile
from importlib.metadata import version
from pathlib import Path

import click
from quantinuum_qircheck import qircheck

from hugr_qir._hugr_qir import cli


@click.command(name="hugr-qir")
@click.argument("hugr_file", type=click.Path(exists=True, path_type=Path))
@click.option(
    "--validate-qir/--no-validate-qir",
    "validate",
    default=True,
    help="Whether to validate output",
)
@click.option(
    "-o",
    "--output",
    "outfile",
    type=click.File("w"),
    default="-",
    help="Name of output file (optional)",
)
@click.version_option(version=version("hugr_qir"))
def hugr_qir(validate: bool, hugr_file: Path, outfile: click.File) -> None:
    """This is the cli for converting hugr to qir."""

    options = ["-q"]
    with tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_file:
        tmp_options = [*options, "-o", temp_file.name]
        cli(str(hugr_file), *tmp_options)
        with Path.open(Path(temp_file.name)) as output:
            qir = output.read()
    if validate:
        qircheck(qir)

    if hasattr(outfile, "write"):
        outfile.write(qir)
    else:
        # this should never happen but the check for write is
        # needed to appease mypy (click doesn't handle typing well)
        err_msg = f"Could not write to object {outfile}"
        raise AttributeError(err_msg)


if __name__ == "__main__":
    hugr_qir()
