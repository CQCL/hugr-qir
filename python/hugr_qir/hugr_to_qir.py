import tempfile
from pathlib import Path

from .cli import hugr_qir_impl


def hugr_to_qir(hugr: bytes, validate_qir: bool = True) -> str:
    """A function for converting hugr to qir."""
    with (
        tempfile.NamedTemporaryFile(delete=True, suffix=".hugr") as temp_infile,
        tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_outfile,
    ):
        with Path.open(Path(temp_infile.name), "wb") as cli_input:
            cli_input.write(hugr)
        with Path.open(Path(temp_outfile.name), "w") as cli_output:
            hugr_qir_impl(validate_qir, Path(cli_input.name), cli_output)
        with Path.open(Path(temp_outfile.name), "r") as cli_output:
            return cli_output.read()
