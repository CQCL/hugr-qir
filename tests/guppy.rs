use std::{env, path::Path, process::{self, Command, Stdio}};

use hugr::{package::Package, std_extensions::STD_REG, Hugr};
use hugr_llvm::inkwell;
use hugr_qir::CompileArgs;
use insta::assert_snapshot;
use itertools::Itertools as _;
use rstest::rstest;

fn capture_guppy(path: impl AsRef<Path>) -> (Hugr, String) {
    let process::Output {
        status,
        stdout,
        stderr,
    } = Command::new(".devenv/state/venv/bin/python")
        .arg(path.as_ref())
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();
    eprintln!("{}", std::str::from_utf8(&stdout).unwrap());
    eprintln!("{}", std::str::from_utf8(&stderr).unwrap());
    assert!(status.success());
    let hugr = Package::from_json_reader(stdout.as_slice(), &STD_REG)
        .unwrap()
        .modules
        .into_iter()
        .exactly_one()
        .unwrap();
    let stderr = String::from_utf8(stderr).unwrap();
    (hugr, stderr)
}

fn compile(hugr: &mut Hugr) -> String {
    let args = {
        let mut a = CompileArgs::default();
        a.validate = true;
        a
    };
    let context = inkwell::context::Context::create();
    let module = args.compile(hugr, &context).unwrap();
    module.to_string()
}

// This generates a test case for each file matching 'guppy_examples/**/*.py'
//
// For now we snapshot test the stderr output of executing the python file,
// and the LLVM IR output by hugr-qir.
//
// If we get problems with new test cases not being run in dev environments we might consider
// adding a `build.rs` as in:
// <https://docs.rs/rstest/latest/rstest/attr.rstest.html#files-path-as-input-arguments>
#[rstest]
fn guppy_examples(
    #[base_dir = "guppy_examples"]
    #[files("**/*.py")]
    file: impl AsRef<Path>,
) {
    let file2 = file.as_ref();
    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file2.file_stem().unwrap().to_str().unwrap());
    settings.bind(|| {
        let (mut hugr, stderr) = capture_guppy(file);
        assert_snapshot!("llvmir", compile(&mut hugr));
        assert_snapshot!("stderr", stderr);
    });
}
