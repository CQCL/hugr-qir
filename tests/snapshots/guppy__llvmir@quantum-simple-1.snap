---
source: tests/guppy.rs
expression: compile(&mut hugr)
---
; ModuleID = 'hugr-qir'
source_filename = "hugr-qir"

%QUBIT = type opaque
%RESULT = type opaque

@0 = private unnamed_addr constant [37 x i8] c"Expected variant 1 but got variant 0\00", align 1
@prelude.panic_template = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@1 = private unnamed_addr constant [2 x i8] c"0\00", align 1

define void @__hugr__.main.1() #0 {
alloca_block:
  %"15_0" = alloca {}, align 8
  %"19_0" = alloca { i1, %QUBIT* }, align 8
  %"35_0" = alloca double, align 8
  %"33_0" = alloca double, align 8
  %"31_0" = alloca double, align 8
  %"20_0" = alloca %QUBIT*, align 8
  %"0" = alloca %QUBIT*, align 8
  %"25_0" = alloca { i32, i8* }, align 8
  %"26_0" = alloca %QUBIT*, align 8
  %"05" = alloca %QUBIT*, align 8
  %"28_0" = alloca %QUBIT*, align 8
  %"36_0" = alloca %QUBIT*, align 8
  %"37_0" = alloca %QUBIT*, align 8
  %"38_0" = alloca i1, align 1
  %"14_0" = alloca {}, align 8
  %"11_0" = alloca {}, align 8
  %"39_0" = alloca i1, align 1
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  br label %0

0:                                                ; preds = %entry_block
  %1 = call %QUBIT* @__quantum__rt__qubit_allocate()
  %2 = insertvalue { i1, %QUBIT* } { i1 true, %QUBIT* poison }, %QUBIT* %1, 1
  store { i1, %QUBIT* } %2, { i1, %QUBIT* }* %"19_0", align 8
  store double 0xBFF921FB54442D18, double* %"35_0", align 8
  store double 0x3FF921FB54442D18, double* %"33_0", align 8
  store double 0x400921FB54442D18, double* %"31_0", align 8
  %"19_01" = load { i1, %QUBIT* }, { i1, %QUBIT* }* %"19_0", align 8
  %3 = extractvalue { i1, %QUBIT* } %"19_01", 0
  switch i1 %3, label %4 [
    i1 true, label %5
  ]

4:                                                ; preds = %0
  br label %cond_20_case_0

5:                                                ; preds = %0
  %6 = extractvalue { i1, %QUBIT* } %"19_01", 1
  store %QUBIT* %6, %QUBIT** %"05", align 8
  br label %cond_20_case_1

7:                                                ; preds = %15
  ret void

cond_20_case_0:                                   ; preds = %4
  store { i32, i8* } { i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @0, i32 0, i32 0) }, { i32, i8* }* %"25_0", align 8
  %"25_03" = load { i32, i8* }, { i32, i8* }* %"25_0", align 8
  %8 = extractvalue { i32, i8* } %"25_03", 0
  %9 = extractvalue { i32, i8* } %"25_03", 1
  %10 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template, i32 0, i32 0), i32 %8, i8* %9)
  call void @abort()
  store %QUBIT* null, %QUBIT** %"26_0", align 8
  %"26_04" = load %QUBIT*, %QUBIT** %"26_0", align 8
  store %QUBIT* %"26_04", %QUBIT** %"0", align 8
  br label %cond_exit_20

cond_20_case_1:                                   ; preds = %5
  %"06" = load %QUBIT*, %QUBIT** %"05", align 8
  store %QUBIT* %"06", %QUBIT** %"28_0", align 8
  %"28_07" = load %QUBIT*, %QUBIT** %"28_0", align 8
  store %QUBIT* %"28_07", %QUBIT** %"0", align 8
  br label %cond_exit_20

cond_exit_20:                                     ; preds = %cond_20_case_1, %cond_20_case_0
  %"02" = load %QUBIT*, %QUBIT** %"0", align 8
  store %QUBIT* %"02", %QUBIT** %"20_0", align 8
  %"20_08" = load %QUBIT*, %QUBIT** %"20_0", align 8
  %"33_09" = load double, double* %"33_0", align 8
  %"35_010" = load double, double* %"35_0", align 8
  call void @__quantum__qis__phasedx__body(double %"33_09", double %"35_010", %QUBIT* %"20_08")
  store %QUBIT* %"20_08", %QUBIT** %"36_0", align 8
  %"36_011" = load %QUBIT*, %QUBIT** %"36_0", align 8
  %"31_012" = load double, double* %"31_0", align 8
  call void @__quantum__qis__rz__body(double %"31_012", %QUBIT* %"36_011")
  store %QUBIT* %"36_011", %QUBIT** %"37_0", align 8
  %"37_013" = load %QUBIT*, %QUBIT** %"37_0", align 8
  %11 = call %RESULT* @__quantum__qis__m__body(%QUBIT* %"37_013")
  %12 = call i1 @__quantum__qis__read_result__body(%RESULT* %11)
  %13 = select i1 %12, i1 true, i1 false
  store i1 %13, i1* %"38_0", align 1
  store {} undef, {}* %"15_0", align 1
  store {} undef, {}* %"14_0", align 1
  store {} undef, {}* %"11_0", align 1
  %"38_014" = load i1, i1* %"38_0", align 1
  %14 = select i1 %"38_014", i1 true, i1 false
  store i1 %14, i1* %"39_0", align 1
  %"39_015" = load i1, i1* %"39_0", align 1
  call void @__quantum__rt__bool_record_output(i1 %"39_015", i8* getelementptr inbounds ([2 x i8], [2 x i8]* @1, i32 0, i32 0))
  %"15_016" = load {}, {}* %"15_0", align 1
  store {} %"15_016", {}* %"15_0", align 1
  %"15_017" = load {}, {}* %"15_0", align 1
  switch i1 false, label %15 [
  ]

15:                                               ; preds = %cond_exit_20
  br label %7
}

declare %QUBIT* @__quantum__rt__qubit_allocate()

declare i32 @printf(i8*, ...)

declare void @abort()

declare void @__quantum__qis__phasedx__body(double, double, %QUBIT*)

declare void @__quantum__qis__rz__body(double, %QUBIT*)

declare %RESULT* @__quantum__qis__m__body(%QUBIT*)

declare i1 @__quantum__qis__read_result__body(%RESULT*)

declare void @__quantum__rt__bool_record_output(i1, i8*)

attributes #0 = { "entry_point" "output_labeling_schema" "qir_profiles"="custom" "required_num_qubits"="unemplemented" "required_num_results"="unemplemented" }
